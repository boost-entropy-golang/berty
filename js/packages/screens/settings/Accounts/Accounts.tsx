import React from 'react'
import { useTranslation } from 'react-i18next'
import { ScrollView, View, TouchableOpacity, Platform } from 'react-native'
import { withInAppNotification } from 'react-native-in-app-notification'
import { useDispatch, useSelector } from 'react-redux'

import beapi from '@berty/api'
import { AccordionV2 } from '@berty/components/Accordion'
import { GenericAvatar } from '@berty/components/avatars'
import { ButtonSettingV2, Section } from '@berty/components/shared-components'
import { UnifiedText } from '@berty/components/shared-components/UnifiedText'
import { useAppDimensions } from '@berty/contexts/app-dimensions.context'
import { useStyles } from '@berty/contexts/styles'
import { useSwitchAccount } from '@berty/hooks'
import { ScreenFC, useNavigation } from '@berty/navigation'
import {
	selectAccounts,
	selectEmbedded,
	selectSelectedAccount,
	setStateOnBoardingReady,
} from '@berty/redux/reducers/ui.reducer'
import { useThemeColor } from '@berty/store'
import {
	exportAccountToFile,
	importAccountFromDocumentPicker,
} from '@berty/utils/accounts/accountBackup'
import { closeAccountWithProgress, refreshAccountList } from '@berty/utils/accounts/accountUtils'
import { pbDateToNum } from '@berty/utils/convert/time'

const AccountButton: React.FC<beapi.account.IAccountMetadata> = ({
	avatarCid,
	publicKey,
	name,
	accountId,
	error,
}) => {
	const colors = useThemeColor()
	const selectedAccount = useSelector(selectSelectedAccount)
	const selected = selectedAccount === accountId
	const { padding, margin } = useStyles()
	const { scaleSize } = useAppDimensions()
	const switchAccount = useSwitchAccount()

	const heightButton = 50

	const [isHandlingPress, setIsHandlingPress] = React.useState(false)
	const handlePress = React.useCallback(async () => {
		if (isHandlingPress) {
			return
		}
		setIsHandlingPress(true)
		if (selectedAccount !== accountId) {
			return switchAccount(accountId || '')
		}
		return
	}, [accountId, switchAccount, isHandlingPress, selectedAccount])

	return (
		<TouchableOpacity
			onPress={handlePress}
			style={[
				padding.left.scale(40),
				{
					height: heightButton * scaleSize,
					backgroundColor: error
						? colors['secondary-text']
						: selected
						? colors['positive-asset']
						: colors['main-background'],
				},
			]}
		>
			<View style={{ flexDirection: 'row', alignItems: 'center' }}>
				<View style={[{ height: heightButton, flexDirection: 'row', alignItems: 'center' }]}>
					<View
						style={{
							flexDirection: 'row',
							alignItems: 'center',
						}}
					>
						<GenericAvatar
							size={30 * scaleSize}
							cid={avatarCid}
							colorSeed={publicKey}
							nameSeed={name}
						/>
					</View>
				</View>
				<View
					style={[
						margin.left.small,
						{ height: heightButton, flexDirection: 'row', alignItems: 'center' },
					]}
				>
					<UnifiedText>{name}</UnifiedText>
				</View>
			</View>
		</TouchableOpacity>
	)
}

export const Accounts: ScreenFC<'Settings.Accounts'> = withInAppNotification(
	({ showNotification }: any) => {
		const { scaleSize } = useAppDimensions()
		const colors = useThemeColor()
		const dispatch = useDispatch()
		const { navigate } = useNavigation()
		const { t }: { t: any } = useTranslation()
		const selectedAccount = useSelector(selectSelectedAccount)
		const accounts = useSelector(selectAccounts)
		const embedded = useSelector(selectEmbedded)

		React.useEffect(() => {
			refreshAccountList(embedded)
		}, [embedded])

		return (
			<View style={{ backgroundColor: colors['secondary-background'], flex: 1 }}>
				<ScrollView
					bounces={false}
					contentContainerStyle={{ paddingBottom: 12 * scaleSize }}
					showsVerticalScrollIndicator={false}
				>
					{Platform.OS !== 'web' && (
						<Section>
							<ButtonSettingV2
								text={t('settings.accounts.backup-button')}
								last
								onPress={async () => {
									try {
										await exportAccountToFile(selectedAccount)
										showNotification({
											title: t('settings.accounts.backup-notif-title'),
											message: t('settings.accounts.backup-notif-desc'),
											additionalProps: { type: 'message' },
										})
									} catch (e) {
										console.warn('account backup failed:', e)
									}
								}}
							/>
						</Section>
					)}
					<Section>
						<AccordionV2 title={t('settings.accounts.accounts-button')}>
							{[...accounts]
								.sort((a, b) => pbDateToNum(a.creationDate) - pbDateToNum(b.creationDate))
								.map(account => {
									return <AccountButton key={account.accountId} {...account} />
								})}
						</AccordionV2>
					</Section>
					<Section>
						<ButtonSettingV2
							text={t('settings.accounts.create-button')}
							onPress={async () => {
								await closeAccountWithProgress(dispatch)
								dispatch(setStateOnBoardingReady())
							}}
							last={Platform.OS === 'web'}
						/>
						{Platform.OS !== 'web' && (
							<ButtonSettingV2
								text={t('settings.accounts.import-button')}
								onPress={async () => await importAccountFromDocumentPicker(embedded)}
								last
							/>
						)}
						{/* <ButtonSettingV2 text={t('settings.accounts.link-button')} disabled last /> */}
					</Section>
					<Section>
						<ButtonSettingV2
							text={t('settings.accounts.delete-button')}
							onPress={() => navigate('Settings.DeleteAccount')}
							last
						/>
					</Section>
				</ScrollView>
			</View>
		)
	},
)
