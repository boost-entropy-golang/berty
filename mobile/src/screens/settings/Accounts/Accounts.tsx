import React from 'react'
import { useTranslation } from 'react-i18next'
import { ScrollView, View, Platform } from 'react-native'
import { withInAppNotification } from 'react-native-in-app-notification'
import { useSelector } from 'react-redux'

import { MenuItem, ItemSection } from '@berty/components'
import { useAppDimensions } from '@berty/contexts/app-dimensions.context'
import { useThemeColor } from '@berty/hooks'
import { ScreenFC, useNavigation } from '@berty/navigation'
import { selectSelectedAccount } from '@berty/redux/reducers/ui.reducer'
import { exportAccountToFile, refreshAccountList } from '@berty/utils/accounts'

export const Accounts: ScreenFC<'Settings.Accounts'> = withInAppNotification(
	({ showNotification }: any) => {
		const { scaleSize } = useAppDimensions()
		const colors = useThemeColor()
		const { navigate } = useNavigation()
		const { t } = useTranslation()
		const selectedAccount = useSelector(selectSelectedAccount)

		React.useEffect(() => {
			refreshAccountList()
		}, [])

		return (
			<View style={{ backgroundColor: colors['secondary-background'], flex: 1 }}>
				<ScrollView
					bounces={false}
					contentContainerStyle={{ paddingBottom: 12 * scaleSize }}
					showsVerticalScrollIndicator={false}
				>
					{Platform.OS !== 'web' && (
						<ItemSection>
							<MenuItem
								onPress={async () => {
									try {
										await exportAccountToFile(selectedAccount)
										showNotification({
											title: t('settings.accounts.backup-notif-title'),
											message: t('settings.accounts.backup-notif-desc'),
											additionalProps: { type: 'message' },
										})
									} catch (e) {
										console.warn('account backup failed:', e)
									}
								}}
							>
								{t('settings.accounts.backup-button')}
							</MenuItem>
						</ItemSection>
					)}
					<ItemSection>
						<MenuItem onPress={() => navigate('Settings.DeleteAccount')}>
							{t('settings.accounts.delete-button')}
						</MenuItem>
					</ItemSection>
				</ScrollView>
			</View>
		)
	},
)
