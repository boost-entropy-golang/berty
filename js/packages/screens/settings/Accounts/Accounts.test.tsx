import { renderScreen } from '@berty/utils/testing/renderScreen.test'

import { Accounts } from './Accounts'

test('Settings.Accounts renders correctly', () => {
	const { toJSON } = renderScreen('Settings.Accounts', Accounts)
	expect(toJSON()).toMatchSnapshot()
})
