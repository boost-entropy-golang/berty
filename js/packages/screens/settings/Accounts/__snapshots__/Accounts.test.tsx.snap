// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Settings.Accounts renders correctly 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "#EDEDED",
          "flex": 1,
        }
      }
    >
      <RCTScrollView
        bounces={false}
        contentContainerStyle={
          Object {
            "paddingBottom": 12,
          }
        }
        showsVerticalScrollIndicator={false}
      >
        <View>
          <View
            style={
              Array [
                Object {
                  "marginTop": 24,
                },
                Object {
                  "marginHorizontal": 16,
                },
                Object {
                  "backgroundColor": "white",
                  "borderRadius": 14,
                },
              ]
            }
          >
            <View
              accessible={true}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "height": 48,
                  "justifyContent": "space-between",
                  "opacity": 1,
                  "padding": 12,
                }
              }
            >
              <View
                style={
                  Object {
                    "width": "75%",
                  }
                }
              >
                <Text
                  numberOfLines={1}
                  style={
                    Array [
                      Object {
                        "fontSize": 30,
                      },
                      Object {
                        "color": "#383B62",
                        "fontFamily": "Open Sans",
                      },
                      Array [
                        Object {
                          "marginLeft": 9,
                        },
                      ],
                    ]
                  }
                >
                  settings.accounts.backup-button
                </Text>
              </View>
              <View
                style={
                  Object {
                    "transform": Array [
                      Object {
                        "scale": 1,
                      },
                    ],
                  }
                }
              >
                <RNSVGMock
                  fill="#393C63"
                  height={20}
                  viewBox="0 0 24 24"
                  width={20}
                />
              </View>
            </View>
          </View>
          <View
            style={
              Array [
                Object {
                  "marginTop": 24,
                },
                Object {
                  "marginHorizontal": 16,
                },
                Object {
                  "backgroundColor": "white",
                  "borderRadius": 14,
                },
              ]
            }
          >
            <View
              accessible={true}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "height": 48,
                  "opacity": 1,
                  "paddingHorizontal": 16,
                }
              }
            >
              <View
                style={
                  Object {
                    "width": "80%",
                  }
                }
              >
                <Text
                  numberOfLines={1}
                  style={
                    Array [
                      Object {
                        "fontSize": 30,
                      },
                      Object {
                        "color": "#383B62",
                        "fontFamily": "Open Sans",
                      },
                      Array [
                        Object {
                          "marginLeft": 9,
                        },
                      ],
                    ]
                  }
                >
                  settings.accounts.accounts-button
                </Text>
              </View>
              <View
                style={
                  Object {
                    "alignItems": "flex-end",
                    "flex": 1,
                  }
                }
              >
                <View
                  style={
                    Object {
                      "transform": Array [
                        Object {
                          "rotate": "0deg",
                        },
                      ],
                    }
                  }
                >
                  <View
                    style={
                      Object {
                        "transform": Array [
                          Object {
                            "scale": 1,
                          },
                        ],
                      }
                    }
                  >
                    <RNSVGMock
                      fill="#393C63"
                      height={20}
                      viewBox="0 0 24 24"
                      width={20}
                    />
                  </View>
                </View>
              </View>
            </View>
            <RCTScrollView
              nestedScrollEnabled={true}
              scrollEventThrottle={0.0001}
              showsVerticalScrollIndicator={false}
              style={
                Object {
                  "borderBottomLeftRadius": 14,
                  "borderBottomRightRadius": 14,
                  "maxHeight": 0,
                }
              }
            >
              <View />
            </RCTScrollView>
          </View>
          <View
            style={
              Array [
                Object {
                  "marginTop": 24,
                },
                Object {
                  "marginHorizontal": 16,
                },
                Object {
                  "backgroundColor": "white",
                  "borderRadius": 14,
                },
              ]
            }
          >
            <View
              accessible={true}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "height": 48,
                  "justifyContent": "space-between",
                  "opacity": 1,
                  "padding": 12,
                }
              }
            >
              <View
                style={
                  Object {
                    "width": "75%",
                  }
                }
              >
                <Text
                  numberOfLines={1}
                  style={
                    Array [
                      Object {
                        "fontSize": 30,
                      },
                      Object {
                        "color": "#383B62",
                        "fontFamily": "Open Sans",
                      },
                      Array [
                        Object {
                          "marginLeft": 9,
                        },
                      ],
                    ]
                  }
                >
                  settings.accounts.create-button
                </Text>
              </View>
              <View
                style={
                  Object {
                    "transform": Array [
                      Object {
                        "scale": 1,
                      },
                    ],
                  }
                }
              >
                <RNSVGMock
                  fill="#393C63"
                  height={20}
                  viewBox="0 0 24 24"
                  width={20}
                />
              </View>
            </View>
            <View
              style={
                Object {
                  "backgroundColor": "#EDF1F7",
                  "height": 1,
                }
              }
            />
            <View
              accessible={true}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "height": 48,
                  "justifyContent": "space-between",
                  "opacity": 1,
                  "padding": 12,
                }
              }
            >
              <View
                style={
                  Object {
                    "width": "75%",
                  }
                }
              >
                <Text
                  numberOfLines={1}
                  style={
                    Array [
                      Object {
                        "fontSize": 30,
                      },
                      Object {
                        "color": "#383B62",
                        "fontFamily": "Open Sans",
                      },
                      Array [
                        Object {
                          "marginLeft": 9,
                        },
                      ],
                    ]
                  }
                >
                  settings.accounts.import-button
                </Text>
              </View>
              <View
                style={
                  Object {
                    "transform": Array [
                      Object {
                        "scale": 1,
                      },
                    ],
                  }
                }
              >
                <RNSVGMock
                  fill="#393C63"
                  height={20}
                  viewBox="0 0 24 24"
                  width={20}
                />
              </View>
            </View>
          </View>
          <View
            style={
              Array [
                Object {
                  "marginTop": 24,
                },
                Object {
                  "marginHorizontal": 16,
                },
                Object {
                  "backgroundColor": "white",
                  "borderRadius": 14,
                },
              ]
            }
          >
            <View
              accessible={true}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "height": 48,
                  "justifyContent": "space-between",
                  "opacity": 1,
                  "padding": 12,
                }
              }
            >
              <View
                style={
                  Object {
                    "width": "75%",
                  }
                }
              >
                <Text
                  numberOfLines={1}
                  style={
                    Array [
                      Object {
                        "fontSize": 30,
                      },
                      Object {
                        "color": "#383B62",
                        "fontFamily": "Open Sans",
                      },
                      Array [
                        Object {
                          "marginLeft": 9,
                        },
                      ],
                    ]
                  }
                >
                  settings.accounts.delete-button
                </Text>
              </View>
              <View
                style={
                  Object {
                    "transform": Array [
                      Object {
                        "scale": 1,
                      },
                    ],
                  }
                }
              >
                <RNSVGMock
                  fill="#393C63"
                  height={20}
                  viewBox="0 0 24 24"
                  width={20}
                />
              </View>
            </View>
          </View>
        </View>
      </RCTScrollView>
    </View>
  </View>
</RNCSafeAreaProvider>
`;
